kmir/src/kmir/smir.py

```diff
@@ -369,3 +379,58 @@ def compute_closure(start: Ty, edges: dict[Ty, set[Ty]]) -> set[Ty]:
             if next in edges:
                 work.extend(edges[next])
     return reached
+
+
+@dataclass
+class AllocInfo:
+    alloc_id: AllocId
+    ty: Ty
+    global_alloc: GlobalAlloc
+
+    @staticmethod
+    def from_dict(dct: dict[str, Any]) -> AllocInfo:
+        return AllocInfo(
+            alloc_id=AllocId(dct['alloc_id']),
+            ty=Ty(dct['ty']),
+            global_alloc=GlobalAlloc.from_dict(dct['global_alloc']),
+        )
+
+
+class GlobalAlloc(ABC):  # noqa: B024
+    @staticmethod
+    def from_dict(dct: dict[str, Any]) -> GlobalAlloc:
+        match dct:
+            case {'Memory': _}:
+                return Memory.from_dict(dct)
+            case _:
+                raise ValueError('Unsupported or invalid GlobalAlloc data: {dct}')
+
+
+@dataclass
+class Memory(GlobalAlloc):
+    allocation: Allocation
+
+    @staticmethod
+    def from_dict(dct: dict[str, Any]) -> Memory:
+        return Memory(
+            allocation=Allocation.from_dict(dct['Memory']),
+        )
+
+
+@dataclass
+class Allocation:
+    data: bytes  # field 'bytes'
+    # provenance
```

Comment:
Any particular reason not to add the provenance here?

Reply:
Added.